src/main/java/io/pravega/perf/PravegaWriterWorker.java

```diff
@@ -52,12 +62,25 @@ public long recordWrite(byte[] data, TriConsumer record) {
         ret.thenAccept(d -> {
             record.accept(time, System.currentTimeMillis(), data.length);
         });
+        noteTimePeriodically();
         return time;
     }
 
     @Override
     public void writeData(byte[] data) {
         producer.writeEvent(data);
+        noteTimePeriodically();
+    }
+
+    private void noteTimePeriodically() {
+        if (writeWatermarkPeriodMillis >= 0) {
+            final long time = System.currentTimeMillis();
+            if (time > nextNoteTime) {
+                producer.noteTime(time);
+                log.debug("noteTimePeriodically: noteTime({})", time);
+                nextNoteTime += writeWatermarkPeriodMillis;
```

Comment:
Is this always in a single thread?

Reply:
Yes. There is always exactly one thread per `PravegaWriterWorker` instance. I added a comment to indicate this.